consensus/src/processes/ghostdag/ordering.rs

```diff
@@ -44,8 +44,16 @@ impl Ord for SortableBlock {
 impl<T: GhostdagStoreReader, S: RelationsStoreReader, U: ReachabilityService, V: HeaderStoreReader> GhostdagManager<T, S, U, V> {
     pub fn sort_blocks(&self, blocks: impl IntoIterator<Item = Hash>) -> Vec<Hash> {
         let mut sorted_blocks: Vec<Hash> = blocks.into_iter().collect();
-        sorted_blocks
-            .sort_by_cached_key(|block| SortableBlock { hash: *block, blue_work: self.ghostdag_store.get_blue_work(*block).unwrap() });
+        sorted_blocks.sort_by_cached_key(|block| SortableBlock {
+            hash: *block,
+            // Since we're only calculating GD at all levels on-demand, we may get blocks from the relations
```

Comment:
This comment is a little out of context because this function doesn't use `relations`. I guess you're referring to the GD calculation in the process of ordering the mergeset? In which part of this process do we rely on `relations`? I thought we use `parents_at_level` for that.

About `make relations store only return parents at the same or higher level`: can you specify what function do you refer that doesn't do it?

Reply:
I encounter a panic when when doing GD in `calc_gd_for_all_levels` during pruning point movement which happens `.unwrap()` that I switched to `.unwrap_or_default()`. I agree it's a little out of context here, but I placed it here so I can pinpoint where exactly the error occurs.

> About make relations store only return parents at the same or higher level: can you specify what function do you refer that doesn't do it?

This function `unordered_mergeset_without_selected_parent`, and specifically this line in `mergeset.rs`:

```
let current_parents = relations.get_parents(current).unwrap();
```

Is where I thought that the relations may be returning parents that are not in the level - but looking at it again, I think this assumption is wrong since `relations` passed to GD manager should already be relation service at a given level.

I need to revisit this change and dig deeper into what actually is causing this code to try to look for hash that isn't there yet (or maybe is no longer there?).

Reply:
I reverted this specific change. It doesn't seem like I'm encountering the panic anymore, likely due to cleaning up a bunch of the ghostdag_store references everywhere.

Reply:
Nvm, I had to put it back as I encountered this panic again when the node was pruning:

```
2024-07-06 12:16:01.923-06:00 [ERROR] thread 'pruning-processor' panicked at consensus/src/processes/ghostdag/ordering.rs:55:66: called `Result::unwrap()` on an `Err` value: KeyNotFound(TempGhostdagCompact/0e005a0e5eaac027ea36de5dc265694273ebeca9b012cb314fe6ea8cd19c87715b02)
thread 'pruning-processor' panicked at consensus/src/processes/ghostdag/ordering.rs:55:66:
called `Result::unwrap()` on an `Err` value: KeyNotFound(TempGhostdagCompact/0e005a0e5eaac027ea36de5dc265694273ebeca9b012cb314fe6ea8cd19c87715b02)
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Exiting...
2024-07-06 12:16:01.954-06:00 [ERROR] thread 'virtual-processor' panicked at consensus/src/pipeline/virtual_processor/processor.rs:431:29: unexpected error rocksdb error Corruption: Mismatch in unique ID on table file 264926. Expected: {7800974614205984544,1475903054101850025} Actual: {6918025063187695999,149668763839083335}
thread 'virtual-processor' panicked at consensus/src/pipeline/virtual_processor/processor.rs:431:29:
unexpected error rocksdb error Corruption: Mismatch in unique ID on table file 264926. Expected: {7800974614205984544,1475903054101850025} Actual: {6918025063187695999,149668763839083335}
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Exiting...
```

This happens during pruning, but I don't know what's causing it exactly yet. I think my original comment about getting parents not in the same level is wrong. Perhaps the issue is that this GD call is trying to get data that has just recently been pruned.

Reply:
In my original comment, this is where I was thinking that the call to `.ghostdag(parents)` eventually will try to get parents which may not be on the same level, and therefore will not exist in the `self.ghostdag_store` of that level: https://github.com/kaspanet/rusty-kaspa/blob/6a56461ff1249e11e3124090253ca0bf128f7189/consensus/src/processes/ghostdag/mergeset.rs#L20

Doing a bit more debugging, this is the error that I'm getting when trying to fill the ghostdag data of level 15 (during build_pruning_proof) in TN11:

```
2024-07-07 12:17:28.418-06:00 [ERROR] thread 'pruning-processor' panicked at consensus/src/processes/ghostdag/ordering.rs:56:17: Failed looking for this block: 97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf
thread 'pruning-processor' panicked at consensus/src/processes/ghostdag/ordering.rs:56:17:
Failed looking for this block: 97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf
```

(the log above is custom, but in master this corresponds to the unwrap at line https://github.com/kaspanet/rusty-kaspa/blob/6a56461ff1249e11e3124090253ca0bf128f7189/consensus/src/processes/ghostdag/ordering.rs#L48)

Now, I tried calculating the current DAG level of this specific hash `97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf` and I got this in the logs:

```
2024-07-07 12:17:28.404-06:00 [WARN ] 97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf => dag level 14
```

So it seems like even though we're processing level 15, we may got parents from different (lower) levels when doing `relations.get_parents()` as in `mergeset.rs` above. Is this expected behavior of `relations.get_parents()`?

Reply:
More debugging details:

Given we are processing level 15, and we note down the dag level of some interesting hashes:

```
2024-07-07 14:03:43.612-06:00 [WARN ] Level: 15 | Root: a5dd609448bcc80b8083194ace12baf512d4c1a6cf8bff3393d55968139eeb1b | Tip: a5efd2ce222f6f3e72316532768587923bdb9c96607ef5cc34bed1001ae658d2
2024-07-07 14:03:43.612-06:00 [WARN ] 2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50 => dag level 14
2024-07-07 14:03:43.612-06:00 [WARN ] 97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf => dag level 14
```

This is the BFS forward iteration from `Root` to `Tip`:
```
2024-07-07 14:03:43.612-06:00 [WARN ] Level: 15 | Hash: a5dd609448bcc80b8083194ace12baf512d4c1a6cf8bff3393d55968139eeb1b
2024-07-07 14:03:43.613-06:00 [WARN ] Level: 15 | Hash: a5dd609448bcc80b8083194ace12baf512d4c1a6cf8bff3393d55968139eeb1b | Child: 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991
2024-07-07 14:03:43.613-06:00 [WARN ] Level: 15 | Hash: a5dd609448bcc80b8083194ace12baf512d4c1a6cf8bff3393d55968139eeb1b | Child: 3be8ced5bca8b9615068bf1e00ab49390598ddd20208870d42a73882dccdcfe9
2024-07-07 14:03:43.613-06:00 [WARN ] Level: 15 | Hash: 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991
2024-07-07 14:03:43.623-06:00 [INFO ] mergeset.rs::unordered_mergeset_without_selected_parent | current: 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991 | relations.get_parents(current): [a5dd609448bcc80b8083194ace12baf512d4c1a6cf8bff3393d55968139eeb1b]

2024-07-07 14:03:43.614-06:00 [WARN ] Level: 15 | Hash: 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991 | Child: f7c78edec9c10204ed8b6827364edd04f835094750f1a123b23a7dede163c580
2024-07-07 14:03:43.614-06:00 [WARN ] Level: 15 | Hash: 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991 | Child: 2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50
2024-07-07 14:03:43.614-06:00 [WARN ] Level: 15 | Hash: 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991 | Child: bc162a2a937e30d5cedd84f8e898e6ac73b4a41d165c8695cf58ac876d86b1a9
2024-07-07 14:03:43.614-06:00 [WARN ] Level: 15 | Hash: 3be8ced5bca8b9615068bf1e00ab49390598ddd20208870d42a73882dccdcfe9
2024-07-07 14:03:43.621-06:00 [WARN ] Level: 15 | Hash: 3be8ced5bca8b9615068bf1e00ab49390598ddd20208870d42a73882dccdcfe9 | Child: 8cf27c51e35bd9227cd20a76d41b5bc3a7f504cdb7032dfca93daaf134303dad
2024-07-07 14:03:43.621-06:00 [WARN ] Level: 15 | Hash: 3be8ced5bca8b9615068bf1e00ab49390598ddd20208870d42a73882dccdcfe9 | Child: bc162a2a937e30d5cedd84f8e898e6ac73b4a41d165c8695cf58ac876d86b1a9
2024-07-07 14:03:43.622-06:00 [WARN ] Level: 15 | Hash: 2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50
2024-07-07 14:03:43.624-06:00 [INFO ] mergeset.rs::unordered_mergeset_without_selected_parent | current: 2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50 | relations.get_parents(current): [97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf, 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991]
2024-07-07 14:03:43.626-06:00 [INFO ] mergeset.rs::unordered_mergeset_without_selected_parent | current: 2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50 | relations.get_parents(current): [97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf, 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991]
2024-07-07 14:03:43.626-06:00 [INFO ] mergeset.rs::unordered_mergeset_without_selected_parent | current: f7c78edec9c10204ed8b6827364edd04f835094750f1a123b23a7dede163c580 | relations.get_parents(current): [97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf, 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991]
2024-07-07 14:03:43.626-06:00 [INFO ] mergeset.rs::unordered_mergeset_without_selected_parent | current: 97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf | relations.get_parents(current): [927dd3dd32a9ceec0ee6b1561771185a14b003e1d97c50f2dc938b4405480056]
2024-07-07 14:03:43.627-06:00 [INFO ] mergeset.rs::unordered_mergeset_without_selected_parent | current: 927dd3dd32a9ceec0ee6b1561771185a14b003e1d97c50f2dc938b4405480056 | relations.get_parents(current): [40b48d2a3af40b10345c921b5e987132275337799bd4d51e6a188f7d6f25815c, 26fb1277ec995718517ac9bdc26d5f39fdbe36c0b7b75de424642144ecbc3f1f, 4c64959e29e632759a7a2b350d00c5086e70df786500295c94f10b6cea506593, 6f1a761ad755d81f76c43f023f6bad7c7775891232098b91e514b4200a26bc98]

2024-07-07 14:03:43.622-06:00 [WARN ] Level: 15 | Hash: 2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50 | Child: 1f11c4acbce3fd01bf786f04232ee4a9ed3b86b64b4576b58d9008951a1dd09c
2024-07-07 14:03:43.622-06:00 [WARN ] Level: 15 | Hash: 2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50 | Child: 0f857297b42847d04dd416af26917979653ef7203d071a28e6f61e7820f777d1
2024-07-07 14:03:43.622-06:00 [WARN ] Level: 15 | Hash: f7c78edec9c10204ed8b6827364edd04f835094750f1a123b23a7dede163c580
2024-07-07 14:03:43.623-06:00 [WARN ] Level: 15 | Hash: f7c78edec9c10204ed8b6827364edd04f835094750f1a123b23a7dede163c580 | Child: 1f11c4acbce3fd01bf786f04232ee4a9ed3b86b64b4576b58d9008951a1dd09c
2024-07-07 14:03:43.623-06:00 [WARN ] Level: 15 | Hash: f7c78edec9c10204ed8b6827364edd04f835094750f1a123b23a7dede163c580 | Child: 0f857297b42847d04dd416af26917979653ef7203d071a28e6f61e7820f777d1
2024-07-07 14:03:43.623-06:00 [WARN ] Level: 15 | Hash: bc162a2a937e30d5cedd84f8e898e6ac73b4a41d165c8695cf58ac876d86b1a9

2024-07-07 14:03:43.623-06:00 [WARN ] Level: 15 | Hash: bc162a2a937e30d5cedd84f8e898e6ac73b4a41d165c8695cf58ac876d86b1a9 | Child: 5e3eaf86fbdb217956c06e6339114599b89d88c3c376a584ce2646f0be29631f
2024-07-07 14:03:43.624-06:00 [WARN ] Level: 15 | Hash: 5e3eaf86fbdb217956c06e6339114599b89d88c3c376a584ce2646f0be29631f
2024-07-07 14:03:43.624-06:00 [WARN ] Level: 15 | Hash: 5e3eaf86fbdb217956c06e6339114599b89d88c3c376a584ce2646f0be29631f | Child: 0f857297b42847d04dd416af26917979653ef7203d071a28e6f61e7820f777d1
2024-07-07 14:03:43.624-06:00 [WARN ] Level: 15 | Hash: 1f11c4acbce3fd01bf786f04232ee4a9ed3b86b64b4576b58d9008951a1dd09c
2024-07-07 14:03:43.625-06:00 [WARN ] Level: 15 | Hash: 1f11c4acbce3fd01bf786f04232ee4a9ed3b86b64b4576b58d9008951a1dd09c | Child: 8cf27c51e35bd9227cd20a76d41b5bc3a7f504cdb7032dfca93daaf134303dad
2024-07-07 14:03:43.625-06:00 [WARN ] Level: 15 | Hash: 1f11c4acbce3fd01bf786f04232ee4a9ed3b86b64b4576b58d9008951a1dd09c | Child: 0cca5923823350455382dd5c451466676c2898d6ef501f23b61190a2548ca97a
2024-07-07 14:03:43.625-06:00 [WARN ] Level: 15 | Hash: 1f11c4acbce3fd01bf786f04232ee4a9ed3b86b64b4576b58d9008951a1dd09c | Child: 5e41a55408e4bfddbbed1cba56850d9b2655590c459814439782844e58411065
2024-07-07 14:03:43.625-06:00 [WARN ] Level: 15 | Hash: 8cf27c51e35bd9227cd20a76d41b5bc3a7f504cdb7032dfca93daaf134303dad
2024-07-07 14:03:43.625-06:00 [INFO ] mergeset.rs::unordered_mergeset_without_selected_parent | current: 3be8ced5bca8b9615068bf1e00ab49390598ddd20208870d42a73882dccdcfe9 | relations.get_parents(current): [a5dd609448bcc80b8083194ace12baf512d4c1a6cf8bff3393d55968139eeb1b]
2024-07-07 14:03:43.626-06:00 [WARN ] Level: 15 | Hash: 8cf27c51e35bd9227cd20a76d41b5bc3a7f504cdb7032dfca93daaf134303dad | Child: dcd74d6bd31f4fecbcda823d2f1ffa43c8d913dccea4d7e55a820f898ef51327
2024-07-07 14:03:43.626-06:00 [WARN ] Level: 15 | Hash: 0f857297b42847d04dd416af26917979653ef7203d071a28e6f61e7820f777d1

2024-07-07 14:03:43.627-06:00 [ERROR] thread 'pruning-processor' panicked at consensus/src/processes/ghostdag/ordering.rs:63:17: Failed to find block: 97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf
thread 'pruning-processor' panicked at consensus/src/processes/ghostdag/ordering.rs:63:17:
Failed to find block: 97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf
```
---

Notably, this happens because `relations.get_parents('2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50')` here returned a parent from level 14:

```
2024-07-07 14:03:43.624-06:00 [INFO ] mergeset.rs::unordered_mergeset_without_selected_parent | current: 2bdf6ef9066a4c5b980f6c53e80a122ea977a79157653b090291d83dc620dc50 | relations.get_parents(current): [97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf, 394ea60e79c41f5e2cc56638df23917a60d7a8d6d52058066a3def49d2e2c991]
```

So when `self.ghostdag_store.get_blue_work(*parent).unwrap()` is called where `parent = '97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf'` this panics because `97fc07b5fdd076c9c467a889a538d9f7d735371d0ecff1cd39b8cc34fcaf3ddf` is not in the current level's GD store as it's a hash from level 14.

Reply:
For now, I've fixed the issue as follows:
- unwrap_or_else -> default to 0 if the hash couldn't be found. I have a comment in the code explaining why this can be fine
- in `ghostdag`, skip hashes in the mergeset that are NOT in the GD store

Can discuss further when this PR is reviewed in detail

Reply:
Your PR at [42a3030](https://github.com/kaspanet/rusty-kaspa/pull/494/commits/42a303026f33beb81e1ff3a280277234957d762a) fixes this

Reply:
Just some logs to prove to myself the above fix was legit:

```
2024-08-27 13:07:00.167-06:00 [WARN ] Processing level: 16 | root: 5f5d89c3a0a89419f364e191986b6e3da2a769d75549401c00c00d471cddc6b6 | tip: 7fb7150cb83fe2135b58b71c4e8d536537d011739d72410cc3e3d73f498c6f54
2024-08-27 13:07:00.167-06:00 [WARN ] Is root ancestor of 7afb0a822d3939fca6ab4631a2d42f7a23da18506f9822dc000ae346fe5ef0ed: false
2024-08-27 13:07:00.168-06:00 [WARN ] Current: a4e1005961cd452069c16afb5849bb249ef4c832a72e9a6004a48d6c7196c317 | Parent: dbb318ffac3cda3b93aa5ecf61fd79fa8fe49d43a0979d152670c6147bb72f3c
2024-08-27 13:07:00.169-06:00 [WARN ] Current: dbb318ffac3cda3b93aa5ecf61fd79fa8fe49d43a0979d152670c6147bb72f3c | Parent: ee97a43b839817962a0a0a55f19e68fe3cd2a018ce3308ff3159263b53119c01
...
2024-08-27 13:07:00.169-06:00 [WARN ] Current: 474098b86913ce5dd1d019dd469c409ed720e67b8a874fed82cd536d18d79f0f | Parent: 7afb0a822d3939fca6ab4631a2d42f7a23da18506f9822dc000ae346fe5ef0ed
2024-08-27 13:07:00.169-06:00 [WARN ] Current: fb46ea0f151d8420ca87526d3a527543537ef3dcf2a309f2343d60d8e88defb4 | Parent: 850307ff2ced095744f656f20e474687de4d477c61f443622c481e9eda0bf3d4
2024-08-27 13:07:00.169-06:00 [WARN ] Current: fb46ea0f151d8420ca87526d3a527543537ef3dcf2a309f2343d60d8e88defb4 | Parent: a2ea43ecffca7d958f9d141190abc5cbdbc8459dd7898d1f60c03b72c0875064
...
2024-08-27 13:07:00.169-06:00 [WARN ] Current: 40de62455e96431302ca610871cb5344ec74ac44625bec240c2cbd3e3b70c1b4 | Parent: f30105d6ae65883fe72bb3f7da12fe02ef08d4581dddc4eb6beee239319916fd
2024-08-27 13:07:00.170-06:00 [ERROR] thread 'tokio-runtime-worker' panicked at consensus/src/processes/ghostdag/ordering.rs:68:66: called `Result::unwrap()` on an `Err` value: KeyNotFound(TempGhostdagCompact/16/007afb0a822d3939fca6ab4631a2d42f7a23da18506f9822dc000ae346fe5ef0ed)
thread 'tokio-runtime-worker' panicked at consensus/src/processes/ghostdag/ordering.rs:68:66:
called `Result::unwrap()` on an `Err` value: KeyNotFound(TempGhostdagCompact/16/007afb0a822d3939fca6ab4631a2d42f7a23da18506f9822dc000ae346fe5ef0ed)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Exiting...
```